RFC: C++ build support

stinos opened the pull request:
These are a number of commits which gradually enable external makefiles (be it user C modules or something else) to have C++ code built into the output just like it's done for C code i.e. add files to SRC_QSTR/SRC_CPP/SRC_MOD_CPP/SRC_USERMOD_CPP which then gets scanned for QSTR and built with CXXFLAGS. Enabled (and tested) on unix and esp32 ports, the former so I can easily run tests, the latter in response to e.g. #6233.
I tried to keep changes minimal and in line with how things are done for the C code. Not too happy with how qstr.i.last is created but since my Makefile skills go no further then copy-paste/adjusting of fragments it's the best I could come up with which works in all cases I tried. Could be easier to e.g. modify makeqstrdefs so it can take multiple files so both PP steps can happen separately, not sure.
Even if not all of it is ok now, would be great to at least have some of the commits merged.

stinos committed:
py/makeqstrdefs.py: Fix beavior when scanning non-C preprocessed files.

When process_file() is passed a preprocessed C++ file for instance it
won't find any lines containing .c files and the last_fname variable
remains None so handle that gracefully.

stinos committed:
py/mkrules.mk: Add target for compiling C++ files.

Move the target from the ESP32 Makefile since that does
what is needed already, but also include files from
user C modules as is done for the C files.

stinos committed:
docs: Fix reference to non-existing Makefile flag.

stinos committed:
py/makeqstrdefs.py: Support preprocessing C++ files for QSTR generation.

When SCR_QSTR contains C++ files they should be preprocessed
with the same compiler flags (CXXFLAGS) as they will be
compiled with, to make sure code scanned for QSTR occurrences
is effectively the code used in the rest of the build.
The 'split SCR_QSTR in .c and .cpp files and process each with
different flags' logic isn't trivial to express in a Makefile
and the existing principle for deciding which files to preprocess
was already rather complicated, so the actual preprocessing is
moved into makeqstrdefs.py completely.

stinos committed:
py/py.mk: Support C++ code for user C modules.

Support C++ code in .cpp files by providing CXX counterparts of
the _USERMOD_ flags we have for C already.  This merely enables
the Makefile of user C modules to use variables specific to C++
compilation, it is still up to each port's main Makefile to also
include these in the build.

stinos committed:
esp32, unix: Support building C++ code.

Support building .cpp files and linking them into the micropython
executable in a way similar to how it is done for .c files.  The main
incentive here is to enable user C modules to use C++ files (which are
put in SRC_MOD_CXX by py.mk) since the core itself does not utilize C++.
However to verify build functionality a unix overage test is added.
The esp32 port already has CXXFLAGS so just add the user modules' flags
to it, for the unix port use a copy of the CFLAGS but strip the ones
which are not usable for C++.

dpgeorge commented on py/mkrules.mk:
```diff
@@ -84,9 +102,20 @@ $(OBJ): | $(HEADER_BUILD)/qstrdefs.generated.h $(HEADER_BUILD)/mpversion.h $(OBJ
 	$(ECHO) "GEN $@"
-	$(Q)$(CPP) $(QSTR_GEN_CFLAGS) $(if $(filter $?,$(QSTR_GLOBAL_DEPENDENCIES)),$^,$(if $?,$?,$^)) >$(HEADER_BUILD)/qstr.i.last
+	$(Q)$(CPP) $(QSTR_GEN_CFLAGS) $(HEADER_BUILD)/noqstr.c $(filter-out %.cpp,$(if $(filter $?,$(QSTR_GLOBAL_DEPENDENCIES)),$^,$(if $?,$?,$^))) >$(HEADER_BUILD)/qstr.i.last
+	$(Q)$(CPP) $(QSTR_GEN_CXXFLAGS) $(HEADER_BUILD)/noqstr.cpp $(filter %.cpp,$(if $(filter $?,$(QSTR_GLOBAL_DEPENDENCIES)),$^,$(if $?,$?,$^))) >>$(HEADER_BUILD)/qstr.i.last
```
If both .c and .cpp go through the same preprocessor, isn't it possible to run just one command for them?  Or is it because of the CXXFLAGS that that doesn't work?

Also, does having both lines here now require a `c++`/`g++` compiler to be installed even if the code is pure C?

stinos replied:
> isn't it possible to run just one command for them?

Had it like that first but that causes 2 problems:
- default CFLAGS usually has -std=gnu11 or so which leads to compiler warning that the flag isn't valid for C++ code
- when a user decides to use CXXFLAGS to enable code which contains QSTRs, preprocessing with CFLAGS wouldn't see that code and miss the QSTRs

> does having both lines here now require a c++/g++ compiler to be installed even if the code is pure C

Yes, but that's also the result of me not knowing how to express in make 'only run this line when there are effectively C++ files'.

dpgeorge replied:
Both of those points could be perhaps addressed by filtering out the required items, eg:
```
$(CPP) $(filter-out -std=%,$(QSTR_GEN_CFLAGS)) $(filter -D%,$(QSTR_GEN_CXXFLAGS)) ...
```
but that seems pretty fragile... 

Two options I can see are:
1. Allow `makeqstrdefs.py` to take multiple `qstr.i.last` files (as you suggested).
2. Add a new sub-command to `makeqstrdefs.py` that does the call to cpp with the correct flags and source files (possibly multiple calls of there are both C and C++ code).  This would eliminate the complex `$(if $?,$?,$^)` stuff and also be useful for potential cmake builds that cannot specify this complex set of dependencies.

stinos replied:
Second option seems the way to go then (shell function could also work I guess) - but just to make sure we're on the same page here: since you mention getting rid of the complex stuff this would be a call like
```
(Q)$(PYTHON) $(PY_SRC)/makeqstrdefs.py pp $(CPP) $(HEADER_BUILD) $(HEADER_BUILD)/qstr.i.last "$?" "$^" "$(QSTR_GLOBAL_DEPENDENCIES)"
```
and makeqstrdefs.py does the 'if any global dependencies changed then PP all, else if there are newer dependencies PP only those, else PP all' logic?

dpgeorge replied:
Yes, exactly like that.

But I also thought of another way, option 3: repeat this rule twice, once for C and C++, then add a new rule to cat the outputs together.  Schematically:
```
$(HEADER_BUILD)/qstr.i.last.c: ...
        # parse C files

$(HEADER_BUILD)/qstr.i.last.cpp: ...
        # parse C++ files

$(HEADER_BUILD)/qstr.i.last: $(HEADER_BUILD)/qstr.i.last.c  $(HEADER_BUILD)/qstr.i.last.cpp
        cat $^ > $@
```
Although that might again need special handling when there are no C++ files... and it won't help with eventual cmake support.  So if you have the time to do option 2 then feel free to attempt it.

stinos replied:
Yes I already tried that option 3 and it has the same problem. And concatenating the 2 files together, of which only 1 might just have been built meaning the other one is an older version is also not ideal.

The Python solution is fairly easy, and looks like it works in all possible situations, however the quoting of arguments proves to be rather non-trivial. If the Makefile has
```
CFLAGS = -DSTM32_HAL_H='<stm32$(MCU_SERIES)xx_hal.h>' -DMP_CONFIGFILE='"mpconfigport_nanbox.h"' -DFFCONF_H=\"lib/oofatfs/ffconf.h\"
```
which are actual flags being used across variuos ports, then passing this as `"$(CFLAGS)"` to Python the according call to `print(sys.argv[1])` shows
```
-DSTM32_HAL_H='<stm32xx_hal.h>' -DMP_CONFIGFILE='mpconfigport_nanbox.h' -DFFCONF_H="lib/oofatfs/ffconf.h"
```
I don't understand where the double quotes around mpconfigport_nanbox.h went, so I see no other option then to change the Makefile (and erase all the single quotes in Python to deal with '<stm32xx_hal.h>'), then turn that into a list of strings to pass to `subprocess.call` which also needs some tricks since it should split on spaces, but not when they are in quotes, and the quotes must be retained (which e.g. `shlext.split` cannot do). Currently everything except the Zephyr minimal port builds though.
Given the above problems I'm thinking there must be an easier way (perhaps having the Makefile generate code fragments which Python can execute?) or I'm missing something.

dpgeorge replied:
It seems to work OK for me if the CFLAGS are not quoted, ie:
```
$(Q)$(PYTHON) $(PY_SRC)/makeqstrdefs.py pp $(CPP) cflags $(QSTR_GEN_CFLAGS) cxxflags $(QSTR_GEN_CXXFLAGS) source $(SRC_QSTR) dependencies $(QSTR_GLOBAL_DEP) ...
```
Since flags always start with `-` and source files always have an extension, it should be possible to separate out those parts based on the sentinels cflags/cxxflags/source/dependencies within the long list of command line args.

stinos replied:
Ok good catch, leaving the quotes passes everything as it should be. Went with the sentinel solution and that works well. 

Added a coverage test; was doubting whether to put the module in examples/ but left it in variants/coverage/ after all, any preference?

dpgeorge replied:
> Added a coverage test; was doubting whether to put the module in examples/ but left it in variants/coverage/ after all, any preference?

I think it'll get lost in `ports/unix/variants/coverage/`.  If it's a pure coverage test then it could be `ports/unix/coveragecpp.cpp` and mirror how `ports/unix/coverage.c` works (or, actually, `coverage.c` could be converted to a user C module...).  Or if it's an example then better to go in `examples/usercmodule/cppexample/`.

Really there are two things to test here: 1) including C++ code in a normal build (eg `ports/unix/coveragecpp.cpp`); 2) making a user C++ module (eg `examples/usercmodule/cppexample/`, which could also sit next to a C user module like `examples/usercmodule/cexample/`).  We could have both, but if you want to limit the scope of this PR then simplest thing to do would be to move your existing example to `examples/`.

stinos replied:
Ok. More than once I found myself going to the C module docs to copy-paste the code there into respective files, so I think it would be good if a complete example is readily available (and mentioned in the docs). Indeed probably best to have separate C and C++ examples. I'll add a `ports/unix/coveragecpp.cpp` as well, good idea.

stinos replied:
@dpgeorge didn't think about the fact that coverag.c and coverage.cpp have the same name apart from the extension, so they cannot both be compiled into coverage.o nor use the same dependency file, etc. Going to take some time to fix that (and keeping ESP32 working, not sure how much it relies on having .o for it's C++ code)

dpgeorge replied:
But can't the source files be `coverage.c` and `coveragecpp.cpp`, so they generate different `.o` files?  We'll just have the restriction that all C and C++ source files must have unique base names, but I think that's a very reasonable restriction.

stinos replied:
Sure, that's definitely a lot easier.

All examples added.

dpgeorge commented on .travis.yml:
```diff
@@ -390,7 +390,7 @@ jobs:
       name: "powerpc port build"
       install:
-        - sudo apt-get install gcc-powerpc64le-linux-gnu libc6-dev-ppc64el-cross
+        - sudo apt-get install gcc-powerpc64le-linux-gnu g++-powerpc64le-linux-gnu libc6-dev-ppc64el-cross
```
Is this needed?

stinos replied:
Without it the error is `gcc: error trying to exec ‘cc1plus’`. But again, if the Makefile target can be rewritten to skip te whole C++ part if there are no C++ files this wouldn't bee needed.

dpgeorge commented:
Looks good in general.  +1

dpgeorge commented:
Would it be worth adding a C++ example module (or similar) to the unix (coverage) build?

stinos commented:
Yes an example would be best. Note there are some quirks like one cannot use `MP_DEFINE_CONST_FUN_OBJ_2` because the expansion is not valid C++. So to keep tests simple I usually have a .c file with `extern` declarations and the module definition, and the .cpp the implements the declared functions in an `extern "C"` scope. I assume that's ok here? None of the alternatives are pretty nor simple.

dpgeorge commented:
> So to keep tests simple I usually have a .c file with `extern` declarations and the module definition, and the .cpp the implements the declared functions in an `extern "C"` scope. I assume that's ok here?

Yes that's what I've also done in projects that use C and C++, so I'd say it's good.

stinos committed:
py: Workaround clang error when building misc.h with C++ compiler.

dpgeorge commented on ports/esp32/Makefile:
```diff
@@ -354,6 +354,9 @@ SRC_C = \
 	$(wildcard $(BOARD_DIR)/*.c) \
 	$(SRC_MOD)
 
+SRC_CPP += \
```
Should this be `SRC_CXX` because (1) it matches `CXXFLAGS`; (2) it won't get confused with $(CPP) which is the preprocessor?  I'm not sure what the general convention is with C++ make variables...

stinos replied:
Me neither, I've seen all kinds of things; but using `CXXFLAGS` is very common so make sense using `CXX` everywhere is for consistency.

Note there are also Makefiles which use one single SRC then split it in %.c and %.cpp. We could do that, but it's a bit strange to have it called SRC_C and then add .cpp files to it? Plus I tried that at one point and quickly ran into problems, don't remember what eactly though. But if you think that's better in the end I can look into it.

dpgeorge replied:
I think having separate `SRC_C` and `SRC_CXX` is fine.

dpgeorge commented on ports/unix/variants/coverage/usercmodule/cppexample/example.cpp:
```diff
@@ -0,0 +1,18 @@
+    #include "py/obj.h"
+}
+
+#include <testcppinc.h>
```
This header is empty, so maybe remove it?  Or put ` extern mp_obj_t cppfunc(mp_obj_t a_obj, mp_obj_t b_obj);` in it so that decl doesn't need to be repeated?

stinos replied:
This was actually just to test if passing `-I$(EXAMPLE_MOD_DIR)` via CXXFLAGS_USERMOD works, but indeed putting the declaration in it would actually put it to use so I'll do that.

stinos committed:
examples: Add user C modules.

Add working example code to provide a starting point for users with
files the can just copy, and include the modules in the coverage test
to verify the complete user C module build functionality.  The cexample
module uses the code originally found in cmodules.rst, which has been
updated to reflect this and partially rewritten with more complete
information.

dpgeorge commented on py/makeqstrdefs.py:
```diff
@@ -19,6 +20,35 @@
 _MODE_COMPRESS = "compress"
 
+# File types which will be scanned.
+_CPPFILE_TYPES = [".cc", ".cpp", ".cxx"]
```
this should probably be `_CXXFILE_TYPES`

stinos replied:
Yes. Although in hindsight, I'm not sure it is really needed to have all of these listed?

dpgeorge replied:
True... it only supports `.cpp` files in the Makefile.  So could just get rid of this variable altogether.  Up to you.

stinos replied:
Removed it to keep things simple. Note I also removed the `.lower()` on the extension since that is not really correct anyway.

dpgeorge commented on py/makeqstrdefs.py:
```diff
@@ -19,6 +20,35 @@
+    csources = []
+    cxxsources = []
+    for source in sources:
+        if os.path.splitext(source.rstrip('" '))[1] in _CPPFILE_TYPES:
```
is this rstrip needed?

stinos replied:
No, leftover from previous version

dpgeorge commented on py/mkrules.mk:
```diff
@@ -79,14 +97,14 @@ $(BUILD)/%.pp: %.c
 $(OBJ): | $(HEADER_BUILD)/qstrdefs.generated.h $(HEADER_BUILD)/mpversion.h $(OBJ_EXTRA_ORDER_DEPS)
 
-# The logic for qstr regeneration is:
+# The logic for qstr regeneration (applied by makeqstr.py) is:
```
"applied by makeqstrdefs.py"

dpgeorge commented on examples/usercmodule/cexample/examplemodule.c:
```diff
@@ -0,0 +1,19 @@
+#include "py/obj.h"
```
IMO it'd make sense to just cut-and-paste the C example from the docs, so it includes comments about each part of the source (rather than have a different example here compared to the docs, which may confuse readers when they expect to find the docs example code in this directory)

stinos replied:
Sure. Questions:
- example starts with 'include required definitions' but only obj.h is really required, can I just drop the others? Or do they serve to indicate what the public API is, in which case the comment would better state that as such?
- name 'cexample' can be kept instead of 'example' (just to match with 'cppexample')?
- add similar comments into the C++ example?

dpgeorge replied:
> example starts with 'include required definitions' but only obj.h is really required, can I just drop the others? Or do they serve to indicate what the public API is, in which case the comment would better state that as such?

True.  The only one that's needed is runtime.h (it's a superset of obj.h b/c it includes obj.h, and adds some useful things like mp_raise etc).  And best practice is to just include runtime.h (and then anything else on an as-needed basis).  So my suggestion would be to just have `runtime.h` and remove the other two from the docs so they match the code example.

> name 'cexample' can be kept instead of 'example' (just to match with 'cppexample')?

Yes.

> add similar comments into the C++ example?

If you have time, would be good.

stinos replied:
Ok, ended up rewriting a large part of the docs because just stripping the code and saying 'look in examples' left it quite incomplete and confusing, imo. @andrewleech you wrote the original, any comments? So basically:
- move code from the docs to actual source code, mostly unchanged apart from the include file
- have the docs refer to the source code
- move the part which talks about the PP define under the 'Compiling the cmodule into MicroPython' section to make it clear 2 build actions are required
- add examples both for building the example modules, and a typical user project directory

dpgeorge commented on examples/usercmodule/cexample/micropython.mk:
```diff
@@ -0,0 +1,2 @@
+CEXAMPLE_MOD_DIR := $(USERMOD_DIR)
```
similarly, this file can be copied verbatim from the docs

dpgeorge commented on ports/unix/variants/coverage/mpconfigvariant.mk:
```diff
@@ -9,9 +9,12 @@ CFLAGS += \
 	-Wold-style-definition -Wpointer-arith -Wshadow -Wuninitialized -Wunused-parameter \
 	-DMICROPY_UNIX_COVERAGE
 
+CFLAGS_EXTRA = -DMODULE_CEXAMPLE_ENABLED=1 -DMODULE_CPPEXAMPLE_ENABLED=1
```
I think this should be `CFLAGS += ...`, so that `CFLAGS_EXTRA` can still be set on the command line to build this variant with additional options

stinos replied:
ok

dpgeorge commented:
This is looking really good, now.  Just a few minor comments.

stinos committed:
py/makeqstrdefs.py: Process C++ files as well.

Preprocessed C++ code isn't different from C code when it
comes to QSTR instances so process it as well.

dpgeorge commented:
Just to give an update: this is currently blocked on `thread/stress_schedule.py` failing, due to a problem with GIL'less threading.  I'm investigating a fix but it may take time, so best way forward might be to adjust that test so it doesn't fail.

dpgeorge commented:
> Just to give an update: this is currently blocked on `thread/stress_schedule.py` failing, due to a problem with GIL'less threading. I'm investigating a fix but it may take time, so best way forward might be to adjust that test so it doesn't fail.

A workaround for this failure is provided by #6574

stinos commented:
Perils of multithreading :)
I can rebase on that commit, to get all checks green here?

dpgeorge commented:
> I can rebase on that commit, to get all checks green here?

Yes, please, that would be a good test (although I did test it locally).

stinos commented:
Looks ok, all checks passed.

andrewleech reviewed (commented):
This looks really great, a really useful extension to the cmodules system! 
The changes are compatible with my existing use case so that works for me.

I've got a couple of minor suggestions for the docs changes, but in general thanks for moving the details into example files, that's really useful.

stinos commented on docs/develop/cmodules.rst:
```diff
@@ -29,7 +30,7 @@ Structure of an external C module
 A MicroPython user C module is a directory with the following files:
 
-* ``*.c`` and/or ``*.h`` source code files for your module.
+* ``*.c`` and/or ``*.cpp`` and/or ``*.h`` source code files for your module.
```
ok

stinos commented on docs/develop/cmodules.rst:
```diff
@@ -44,137 +45,126 @@ A MicroPython user C module is a directory with the following files:
   eg ``EXAMPLE_MOD_DIR := $(USERMOD_DIR)``
 
-  Your ``micropython.mk`` must add your modules C files relative to your
+  Your ``micropython.mk`` must add your module's source files relative to your
```
yes

stinos commented on docs/develop/cmodules.rst:
```diff
@@ -44,137 +45,126 @@ A MicroPython user C module is a directory with the following files:
 
   If you have custom ``CFLAGS`` settings or include folders to define, these
-  should be added to ``CFLAGS_USERMOD``.
+  should be added to ``CFLAGS_USERMOD``, or ``CXXFLAGS_USERMOD``.
```
I reworded this part now, using the same terminology as used in the gcc manuals i.e. not 'CFLAGS settings' but 'compiler options' etc.

stinos commented on docs/develop/cmodules.rst:
```diff
@@ -44,137 +45,126 @@ A MicroPython user C module is a directory with the following files:
-``micropython.mk``
+- all modules found in this directory will be compiled, but only those
+  which are explicitly enabled will be availabe for importing. Enabling a
+  module is done by setting the preprocessor define from it's module
```
Yes. Can't believe I keep on making this mistake, I know the rules, yet when typing I forget :)

andrewleech replied:
Oh I make the same sort of mistakes myself all the time - that's why it's great if you can have someone else proof read :-)

stinos replied:
Exactly, and thanks for that!

dpgeorge commented:
Merged!  Thanks for the work on this!

(There are a few minor things I want to follow up with.)

dpgeorge commented:
Merged in 1b723937e3fe4012ac878ded833ff6a39455e0b8 through 25c4563f26c2270cde01b01dca0e8b801c9c8282

dpgeorge commented:
See follow up in #6580
